File: java/base/src/main/java/vsdk/toolkit/raycasting/photonMap/DiscreteSampling.java
package vsdk.toolkit.raycasting.photonMap;

public class DiscreteSampling {
    public static int
    sample(
        double[] probabilities,
        double total,
        double[] x1,
        double[] probabilityDensityFunction)
    {
        int i = 0;
        double sum;
        double left;
        double sample = x1[0] * total;

        sum = probabilities[0];

        while ( sample > sum ) {
            i++;
            sum += probabilities[i];
        }

        // Rescale x_1
        left = sum - probabilities[i];

        x1[0] = ((sample - left) / (sum - left));
        probabilityDensityFunction[0] = probabilities[i] / total;
        return i;
    }
}

File: typescript/base/src/main/vsdk/toolkit/raycasting/photonMap/DiscreteSampling.ts
export class DiscreteSampling {
  private constructor() {
  }

  public static sample(
    probabilities: number[],
    total: number,
    x1: number[],
    probabilityDensityFunction: number[]
  ): number {
    let i = 0;
    let sample = x1[0]! * total;
    let sum = probabilities[0]!;

    while (sample > sum) {
      i++;
      sum += probabilities[i]!;
    }

    // Rescale x_1
    const left = sum - probabilities[i]!;

    x1[0] = ((sample - left) / (sum - left));
    probabilityDensityFunction[0] = probabilities[i]! / total;
    return i;
  }
}
